### PR TITLE
Fix Security's expression resolver to not remove unavailable but authorized names

### LINKED ISSUE #90215: Get Index API returns no indices for a brief amount of time until indices are recovered into cluster_state
Tested on `8.5.0-SNAPSHOT` but I've also observed this behaviour on 8.4.1 clusters on Cloud.

## 1. Create a lot of indices
```
curl -s -k -X PUT "http://elastic:changeme@localhost:9200/test[1-500]" -H 'Content-Type: application/json'
```

## 2. Create a service token (seems to avoid 401 errors while .security is being restored?)
```
bin/elasticsearch-service-tokens create elastic/kibana mytoken
```

## 3. Start Elasticsearch and start polling the Get Index API:
```
for i in {1..10000}
do
    curl -k -w "\n" -X GET "http://localhost:9200/.kibana,.kibana_8.4.1_001?ignore_unavailable=true&features=aliases" -H 'Content-Type: application/json' -H 'Authorization: Bearer your_service_token'
done
```

## Results:
```
curl: (7) Failed to connect to localhost port 9200 after 2 ms: Connection refused

...

curl: (7) Failed to connect to localhost port 9200 after 3 ms: Connection refused

curl: (7) Failed to connect to localhost port 9200 after 3 ms: Connection refused

{}        // <-- One response suggesting .kibana_8.5.0_001 does not exist
{".kibana_8.5.0_001":{"aliases":{".kibana":{"is_hidden":true},".kibana_8.5.0":{"is_hidden":true}},"mappings":{},"settings":{}}}
{".kibana_8.5.0_001":{"aliases":{".kibana":{"is_hidden":true},".kibana_8.5.0":{"is_hidden":true}},"mappings":{},"settings":{}}}
...
```

When Kibana is started at the same time as Elasticsearch, this behaviour causes Kibana to believe that there isn't any existing indices to migrate from. Kibana therefore attempts to initialize the cluster as a new cluster which can lead to a failed upgrade migration and in some circumstances data loss.

### PR DESCRIPTION
**Pull request by albertzaharovits**
For explicit (non-wildcard) names in expressions from requests with the `ignoreUnavailable` request option set to `true`, the Security filter removes names for "unavailable" resources. This behavior is theoretically correct, but in practice the cluster state might not be recovered at the point in time when the Security filter does the rewrite (this causes problems, see #90215); In any case, the logic to remove "unavailable" names is unnecessarily duplicated from Core's expression resolver.

This PR makes the expression resolver in the Security filter to NOT:

1) remove missing but authorized names (indices, aliases, datastreams)
2) remove authorized datastreams if context disallows datastreams
3) remove backing indices of authorized datastreams if context disallows datastreams

It will only remove names of unauthorized resources, and let the unavailable names go through to be handled (to be ignored or throw "not found" exception) by the expression resolver in Core.

Fixes: https://github.com/elastic/elasticsearch/issues/90215

### CODE REVIEW

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authz/IndexAliasesTests.java [hunk 1]
```diff
@@ -769,14 +769,16 @@ public void testGetAliasesAliasesOnlyPermissionIgnoreUnavailable() {
...
+        // no manage_aliases privilege on `non_authorized` alias and ignore_unavailable doesn't influence aliases
+        assertThrowsAuthorizationException(
+            client.admin()
+                .indices()
+                .prepareGetAliases("non_authorized")
+                .addIndices("test_1")
+                .setIndicesOptions(IndicesOptions.lenientExpandOpen())::get,
+            GetAliasesAction.NAME,
+            "aliases_only"
+        );
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authz/IndexAliasesTests.java:781]: The `ignore_unavailable` request option does not currently work for "unavailable" aliases, see: https://github.com/elastic/elasticsearch/blob/844bb4cee573b4eb2b66fd2253d364ae7cf935fa/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java#L375 .

It worked in this particular test by coincidence.
In this particular test the `test_1` index is authorized but non-existent, so the request was previously authorized after the request was rewritten to empty. But now this is not the case anymore.
`test_1` is authorized, so it is not considered "unavailable" from Security's POV,

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 2]
```diff
@@ -880,19 +881,14 @@ static AuthorizedIndices resolveAuthorizedIndicesFromRole(
...
-                if (indexAbstraction.getParentDataStream() != null && predicate.test(indexAbstraction.getParentDataStream())) {
-                    return true;
-                } else {
-                    return predicate.test(indexAbstraction);
-                }
-            } else {
-                return indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM && predicate.test(indexAbstraction);
+                return (indexAbstraction.getParentDataStream() != null && predicate.test(indexAbstraction.getParentDataStream()))
+                    || predicate.test(indexAbstraction);
             }
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:892]: This is the core of the change proposed by this PR.

This predicate here is used to remove explicit names when rewriting the request if `ignore_unavailable` is `true`, see: https://github.com/elastic/elasticsearch/blob/844bb4cee573b4eb2b66fd2253d364ae7cf935fa/server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java#L112

After this expression rewrite, the request goes into the authz stage, where every expression item is checked again: 
https://github.com/elastic/elasticsearch/blob/844bb4cee573b4eb2b66fd2253d364ae7cf935fa/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java#L510

The intent is that the test here, ie:
``` 
final IndexAbstraction indexAbstraction = lookup.get(name);
if (indexAbstraction == null) {
    // missing but authorized resources should be handled downstream in the action handler, not here in the Security filter
    return predicate.testMissingResource(name);
} else {
    // We check the parent data stream first if there is one. For testing requested indices, this is most likely
    // more efficient than checking the index name first because we recommend grant privileges over data stream
    // instead of backing indices.
    return (indexAbstraction.getParentDataStream() != null && predicate.test(indexAbstraction.getParentDataStream()))
        || predicate.test(indexAbstraction);
}
```
be equivalent to the test in `IndicesPermission#isActionGranted`, such that **only unauthorized names be removed when rewriting the expression with ignore_unavailable=true**.

File: x-pack/plugin/eql/qa/security/src/javaRestTest/java/org/elasticsearch/xpack/eql/EqlRestValidationIT.java [hunk 3]
```diff
@@ -23,6 +23,11 @@ protected Settings restClientSettings() {
 
     @Override
     protected String getInexistentIndexErrorMessage() {
+        return "\"root_cause\":[{\"type\":\"verification_exception\",\"reason\":\"Found 1 problem\\nline -1:-1: Unknown index ";
+    }
+
+    @Override
+    protected String getInexistentWildcardErrorMessage() {
```
Comment (albertzaharovits) [inline x-pack/plugin/eql/qa/security/src/javaRestTest/java/org/elasticsearch/xpack/eql/EqlRestValidationIT.java:30]: Some EQL tests assert the audit log messages, and some will change because the transport request, which gets audited, is not rewritten anymore to erase authorized but unavailable indices.

File: x-pack/plugin/eql/qa/rest/src/javaRestTest/java/org/elasticsearch/xpack/eql/EqlRestValidationIT.java [hunk 4]
```diff
@@ -31,6 +31,8 @@ protected void assertErrorMessageWhenAllowNoIndicesIsFalse(String reqParameter)
         // https://github.com/elastic/elasticsearch/issues/64197
         // is closed
         assertErrorMessage("inexistent1,inexistent2", reqParameter, """
-            "root_cause":[{"type":"index_not_found_exception","reason":"no such index [null]\"""");
+            "root_cause":[{"type":"index_not_found_exception","reason":"no such index [null]",
+            "resource.type":"index_expression","resource.id":["inexistent1","inexistent2"]}]
+            """);
```
Comment (albertzaharovits) [inline x-pack/plugin/eql/qa/rest/src/javaRestTest/java/org/elasticsearch/xpack/eql/EqlRestValidationIT.java:36]: Core now throws the "not found" exception, rather than the security code, and it does it in a peculiar way:
https://github.com/elastic/elasticsearch/blob/bbb557b2a15f6340c74644785c6f78cf149e6977/server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java#L453

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java [hunk 5]
```diff
@@ -110,6 +110,8 @@ && isIndexVisible(
                 if (minus) {
                     finalIndices.remove(indexAbstraction);
                 } else if (indicesOptions.ignoreUnavailable() == false || isAuthorized.test(indexAbstraction)) {
+                    // only unauthorized names are considered "unavailable" at this stage
```
Comment (jakelandis) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java:113]: I find these comments really hard to parse... is there any way to simplify either the code or the comments to make this more readable ? I think the biggest issue is the double (tripple?) negatives with a conflation of  terms... for  example "I do not want to ignore expressions that are not available"  is already difficult to parse and introducing treating unauthorized as unavailble reads "I want to treat unauthorized as unavailable which means that if I want to ignore expressions that not available ... I'm too lost to finish ".   

The question I want answered here is why do we care about unavailble at all here ?  this is the security code so it makes to test if authorized, but if the core code will remove unavailable as needed then why do i need to even look at the ignore flag here?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java [hunk 6]
```diff
@@ -159,10 +159,55 @@ private Predicate<IndexAbstraction> buildIndexMatcherPredicateForAction(String a
...
+        public IsResourceAuthorizedPredicate(StringMatcher resourceNameMatcher, StringMatcher additionalNonDatastreamNameMatcher) {
+            this.resourceNameMatcher = resourceNameMatcher;
+            this.additionalNonDatastreamNameMatcher = additionalNonDatastreamNameMatcher;
+        }
+
+        /**
+         * Authorization test for resources.
+         * An additional predicate tests datastreams and backing indices, apart from the regular predicate that tests all resource types.
+         */
+        public final boolean test(IndexAbstraction indexAbstraction) {
```
Comment (jakelandis) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java:184]: for public methods that return boolean values, can you include "returns true when X , false when Y (or false otherwise)"

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 7]
```diff
@@ -880,19 +881,14 @@ static AuthorizedIndices resolveAuthorizedIndicesFromRole(
         }, name -> {
             final IndexAbstraction indexAbstraction = lookup.get(name);
             if (indexAbstraction == null) {
-                return false;
-            }
-            if (includeDataStreams) {
+                // missing but authorized resources should be handled downstream in the action handler, not here in the Security filter
+                return predicate.testMissingResource(name);
```
Comment (jakelandis) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:885]: "missing but authorized resources should be handled downstream..." ... so why is this security code testing for missing ?

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authz/IndexAliasesTests.java (hunk 1 above)
Comment (ywangd) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authz/IndexAliasesTests.java:781]: From end-user's perspective, strictly speaking, this is a slight behaviour change:
* Status code from 404 to 403
* Response content from containing found aliases (and missing aliases) to unauthorized error response

Just want to call it out explicitly. But I don't think it is a real concern because the existing response is not a 200 anyway and it feels like an edge case.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 8]
```diff
@@ -880,19 +881,14 @@ static AuthorizedIndices resolveAuthorizedIndicesFromRole(
         }, name -> {
             final IndexAbstraction indexAbstraction = lookup.get(name);
             if (indexAbstraction == null) {
-                return false;
-            }
-            if (includeDataStreams) {
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:885]: Just to confirm that this is same to remove because core handles whether data stream is allowed in the context? Do we really need to remove it here? It does not seem to be strictly related to the original issue. It is also a bit inconsistent that `includeDataStreams` is still handled in the first lambda (the one backs `AuthorizedIndices.all`) but not here? Or should we also remove logic for `includeDataStreams` in the first lambda as well?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 2 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:892]: Can we unify the loigc of `test` and `testMissingResource` into a single method? It is better that callers do not have to worry about deciding on which method to use. It should be possible if the predicate takes an object that encapsulate both `name` and `indexAbstraction`, something similar to `IndicesPermission#IndexResource`. 

I think ideally it would be great to reuse the existing `IndexResource` and unify the logic between this method and `isActionGranted`. It will likely help understand this part of code. But it does not seem to be an easy job. So we can tackle it later gradually. But I think we can avoid having both `test` and `testMissingResource` as suggest earlier. It also helps the `IsResourceAuthorizedPredicate` to be just a simple `Predicate` and avoid having to implement its own `and` method.

**Review by ywangd — COMMENTED** [review]
I have one main comment around how to structure the new IsResourceAuthorizedPredicate class. Thanks!

**Comment by albertzaharovits** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authz/IndexAliasesTests.java (hunk 1 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authz/IndexAliasesTests.java:781]: Yes, it's indeed a behavior change for an edge case.

Rereading this now, after a week, I realise I haven't done a good job explaining what's going on here.
Let me try again:

The get aliases action in this test uses the "lenient expand open" request option, which means that missing explicit indices are ignored (and that wildcards can expand to empty).
The `test_1` index does not exist, and it is the only index in the request, so when retrieving the `non_authorized` alias for it, because of the request option, this previously returned a 200 empty result, because there were ultimately no indices to return the aliases for.
But, after the change, Security does not care if `test_1` exists or not (Core does, and if Security lets the request through, given the request option, it will return the empty result). For Security, the `test_1` is authorized, but the alias `non_authorized` is not, hence it will then return a 403 because of the alias. For **indices**, the request option `ignore_unavailable` silently discards unauthorized index names, but the option never discards unauthorized **aliases** (but maybe it should?).

Therefore, clients will notice behavior differences for this API, when **all** the indices are authorized but missing, the request option tells to ignore missing indices, and there are unauthorized aliases (missing or not). Before they got a 200 empty, after they get a 403.
They got

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 2 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:892]: There's a subtle problem with making the `IsResourceAuthorizedPredicate` take an `IndexResource` parameter.
The problem is that, when the resource exists, it matters if the "is resource authorized" predicate is called in the context of wildcard expansion, or in the context of checking explicit names.

The difference is that the predicate in the context of checking explicit names first tests on the parent datastream (if applicable) and then it tests on the resource itself. This is not the case for the predicate in the wildcard expansion context.

So, in essence, the `IsResourceAuthorizedPredicate` would still have to implement two slightly different test methods.

Not exactly what you asked for, but I could be hidding this whole
```
final IndexAbstraction indexAbstraction = lookup.get(name);
if (indexAbstraction == null) {
    // missing but authorized resources should be handled downstream in the action handler, not here in the Security filter
    return predicate.testMissingResource(name);
} else {
    // We check the parent data stream first if there is one. For testing requested indices, this is most likely
    // more efficient than checking the index name first because we recommend grant privileges over data stream
    // instead of backing indices.
    return (indexAbstraction.getParentDataStream() != null && predicate.test(indexAbstraction.getParentDataStream()))
        || predicate.test(indexAbstraction);
}
```
block inside a method in the `IsResourceAuthorizedPredicate` class, something of the sorts of `#testExplicitName`. This should be equivalent (and someday replace) the `isActionGranted` method.
What do you think?

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java (hunk 5 above)
Comment (albertzaharovits) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java:113]: The code comment that I added here is more confusing than helpful, apologies.

I changed the comment to:
```
// Unauthorized names are considered unavailable, so if `ignoreUnavailable` is `true` they should be silently
// discarded from the `finalIndices` list. Other "ways of unavailable" must be handled by the action
// handler, see: https://github.com/elastic/elasticsearch/issues/90215
```
Hope this makes it clearer.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 7 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:885]: I've changed the code comment to:
```
// test access (by name) to a resource that does not currently exist
// the action handler must handle the case of accessing resources that do not exist
```
The original comment could only be understood in the context of the change, as a GH comment rather than in the code, apologies.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 8 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:885]: > Just to confirm that this is same to remove because core handles whether data stream is allowed in the context? 

Yes. In Core, accessing a datastream in a context that doesn't allow datastreams is handled similarly to a missing index, or to an alias when the request options ignore aliases.

> Do we really need to remove it here? It does not seem to be strictly related to the original issue. 
On the contrary, I think it would be inconsistent to continue to deal with non-authz unavailable conditions.

> It is also a bit inconsistent that includeDataStreams is still handled in the first lambda (the one backs AuthorizedIndices.all) but not here? Or should we also remove logic for includeDataStreams in the first lambda as well?

Yes, I agree it looks inconsistent. But I don't think we can get away from it. When expanding wildcards we need to know if to expand to datastreams and/or aliases too, but for explicit names we don't (and shouldn't).

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java [hunk 9]
```diff
@@ -159,10 +159,55 @@ private Predicate<IndexAbstraction> buildIndexMatcherPredicateForAction(String a
...
+         * Given another {@link IsResourceAuthorizedPredicate}, return a new one that is equivalent to the conjunction of that other one
+         * and this current one.
+         */
+        public final IsResourceAuthorizedPredicate and(IsResourceAuthorizedPredicate other) {
+            StringMatcher andedResourceNameMatcher = this.resourceNameMatcher.and(other.resourceNameMatcher);
+            StringMatcher andedAdditionalNonDatastreamNameMatcher = (this.resourceNameMatcher.and(other.additionalNonDatastreamNameMatcher))
+                .or((this.additionalNonDatastreamNameMatcher.and(other.resourceNameMatcher)))
+                .or((this.additionalNonDatastreamNameMatcher.and(other.additionalNonDatastreamNameMatcher)));
+            return new IsResourceAuthorizedPredicate(andedResourceNameMatcher, andedAdditionalNonDatastreamNameMatcher);
+        }
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java:206]: I tried to avoid to create another `IndexResource` type of object that needs to wrap every `IndexAbstraction` when expanding wildcards. Without it there needs to be 2 test methods or a test method with 2 parameters, hence the current design.

**Comment by albertzaharovits** [comment]
@jakelandis and @ywangd Thank you for the review, I acknowledge it must've been laborious.
I answered all your points, but I haven't pushed any code changes.
@ywangd We probably need to sync on the `IsResourceAuthorizedPredicate`'s interface, because I'm undecided on the approach, but we can also figure it out over GH too.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authz/IndexAliasesTests.java (hunk 1 above)
Comment (ywangd) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authz/IndexAliasesTests.java:781]: >  Before they got a 200 empty, after they get a 403.

Does it return 200 today? At REST layer, the response code is set to 404 if there is any explicitly requested alias not found.

https://github.com/elastic/elasticsearch/blob/c0e702c8365a068a8e0ff7290264e622158dea03/server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetAliasesAction.java#L130-L133

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 2 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java]: My preference is to avoid having a second test method for `IsResourceAuthorizedPredicate`. It is surprising that a predicate has more than one test method and you need some effort in deciding which one to use. 

Based on your comment here https://github.com/elastic/elasticsearch/pull/92625#discussion_r1071358409

Can we make `IsResourceAuthorizedPredicate` implement `BiPredicate` if you prefer not having a `IndexResource` like class?

Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java]: As we've discussed earlier today, I've reimplemented the `IsResourceAuthorizedPredicate` as a `BiPredicate<String, IndexAbstraction>`, see [878ba9e](https://github.com/elastic/elasticsearch/pull/92625/commits/878ba9e21f5d07ae3fd4a060617b31b532d78d47) :
```
    public static class IsResourceAuthorizedPredicate implements BiPredicate<String, IndexAbstraction> {

        private final BiPredicate<String, IndexAbstraction> biPredicate;

        // public for tests
        public IsResourceAuthorizedPredicate(StringMatcher resourceNameMatcher, StringMatcher additionalNonDatastreamNameMatcher) {
            this((String name, @Nullable IndexAbstraction indexAbstraction) -> {
                assert indexAbstraction == null || name.equals(indexAbstraction.getName());
                return resourceNameMatcher.test(name)
                    || (isPartOfDatastream(indexAbstraction) == false && additionalNonDatastreamNameMatcher.test(name));
            });
        }

        private IsResourceAuthorizedPredicate(BiPredicate<String, IndexAbstraction> biPredicate) {
            this.biPredicate = biPredicate;
        }

        public final IsResourceAuthorizedPredicate and(IsResourceAuthorizedPredicate other) {
            return new IsResourceAuthorizedPredicate(this.biPredicate.and(other.biPredicate));
        }

        public final boolean test(IndexAbstraction indexAbstraction) {
            return test(indexAbstraction.getName(), indexAbstraction);
        }

        @Override
        public boolean test(String name, @Nullable IndexAbstraction indexAbstraction) {
            return biPredicate.test(name, indexAbstraction);
        }

        private static boolean isPartOfDatastream(IndexAbstraction indexAbstraction) {
            return indexAbstraction != null
                && (indexAbstraction.getType() == IndexAbstraction.Type.DATA_STREAM || indexAbstraction.getParentDataStream() != null);
        }
    }
```
For convenience, I also added a `test(IndexAbstraction indexAbstraction)` in addition to `test(String name, @Nullable IndexAbstraction indexAbstraction)`, to avoid the `test(index.getName(), index)` pattern.

Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java]: However, I'm hesitant to change the predicate further, in order to move the logic that tests the parent datastream of a backing index. I still can't see any serious issues, but I very much prefer that such a change be in a separate PR.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java]: We met to discuss this and we landed on the authz predicate with a single test method with to parameters.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 8 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:885]: We met today and also discussed this.
> It is also a bit inconsistent that includeDataStreams is still handled in the first lambda (the one backs AuthorizedIndices.all) but not here? Or should we also remove logic for includeDataStreams in the first lambda as well?

I think we're on the same page that it would be preferable to not check `includeDAtaStreams` when building the result of `AuthorizedIndices.all`. I think the code should work. But It's not obvious if it can affect performance. Therefore I prefer to keep this for a follow-up PR.

**Comment by albertzaharovits** [comment]
@elasticmachine update branch

**Comment by albertzaharovits** [comment]
@elasticmachine update branch

**Review by jakelandis — APPROVED** [review]
LGTM , but would appreciate a second set of eyes for final approval.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java [hunk 10]
```diff
@@ -2175,7 +2182,6 @@ public void testDataStreamNotAuthorizedWhenBackingIndicesAreAuthorizedViaWildcar
         assertThat(authorizedIndices.check("logs-foobar"), is(false));
         DataStream dataStream = metadata.dataStreams().get("logs-foobar");
         assertThat(authorizedIndices.all().get(), not(hasItem(indexName)));
-        assertThat(authorizedIndices.check(indexName), is(false));
```
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java:2178]: This is an interesting point. The indexName `.ds-logs-foobar-*` contains a wildcard. I think this `check` will now return `true` here since the new name check should make it pass. It previously returns `false` because the name does not have an IndexAbstraction. If we believe the new behaviour is correct (returning `true`), should we retain this assertion but change it to `is(true)` and add a comment to say why it is the case? Just deleting it feels like we are leaving it unspecified and I don't think we want that.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java [hunk 11]
```diff
@@ -2237,7 +2242,8 @@ public void testDataStreamNotAuthorizedWhenBackingIndicesAreAuthorizedViaWildcar
...
+        // request pattern is subset of the authorized pattern
+        assertThat(authorizedIndices.check(indexName), is(true));
```
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java:2246]: I see we actually changing the assertion here. This is what I was suggesting above.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/accesscontrol/IndicesPermissionTests.java [hunk 12]
```diff
@@ -663,6 +670,95 @@ public void testIndicesPermissionHasFieldOrDocumentLevelSecurity() {
...
+                    Settings.builder()
+                        .put(SETTING_NUMBER_OF_SHARDS, 1)
+                        .put(SETTING_NUMBER_OF_REPLICAS, 0)
+                        .put(SETTING_INDEX_VERSION_CREATED.getKey(), Version.CURRENT)
+                )
+                .build()
+        );
+        IndicesPermission.IsResourceAuthorizedPredicate predicate = predicate1.and(predicate2);
+        assertThat(predicate.test(dataStreamA), is(false));
+        assertThat(predicate.test(concreteIndexA), is(true));
```
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/accesscontrol/IndicesPermissionTests.java:759]: Nit: Can we add a few more assertions here for the more common cases of the `and` usage? e.g.:
1. resource is allowed by both the predicates with just the resourceNameMatcher
2. resource is allowed by the original predicate but *not* allowed by the limited predicate
3. The opposite of 2
4. Resource is denied by both (This is already covered by the test of `dataStreamA`)

**Review by ywangd — APPROVED** [review]
LGTM

A few minor comments for your consideration.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java [hunk 13]
```diff
@@ -160,10 +161,63 @@ private Predicate<IndexAbstraction> buildIndexMatcherPredicateForAction(String a
...
+
+        /**
+        * Given another {@link IsResourceAuthorizedPredicate} instance in {@param other},
+        * return a new {@link IsResourceAuthorizedPredicate} instance that is equivalent to the conjunction of
+        * authorization tests of that other instance and this one.
+        */
+        public final IsResourceAuthorizedPredicate and(IsResourceAuthorizedPredicate other) {
+            return new IsResourceAuthorizedPredicate(this.biPredicate.and(other.biPredicate));
+        }
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java:195]: Great catch!
I used something simpler, that I think works equally well.
```
@Override
public final IsResourceAuthorizedPredicate and(BiPredicate<? super String, ? super IndexAbstraction> other) {
    return new IsResourceAuthorizedPredicate(this.biPredicate.and(other));
}
```

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java [hunk 14]
```diff
@@ -1071,31 +1065,46 @@ public void testResolveGetAliasesRequestMissingIndexStrict() {
         request.aliases("alias2");
         final AuthorizedIndices authorizedIndices = buildAuthorizedIndices(user, GetAliasesAction.NAME);
         List<String> indices = resolveIndices(request, authorizedIndices).getLocal();
-        // the union of all indices and aliases gets returned, missing is not an existing index/alias but that doesn't make any difference
-        String[] expectedIndices = new String[] { "alias2", "missing" };
-        assertThat(indices, hasSize(expectedIndices.length));
-        assertThat(indices, hasItems(expectedIndices));
+        // explicit names are not erased from the request if `ignoreUnavailable` is `false`
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java]: [a4eb493](https://github.com/elastic/elasticsearch/pull/92625/commits/a4eb493001deaa2f891e664928b595375311bca2)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java]: [ace99f9](https://github.com/elastic/elasticsearch/pull/92625/commits/ace99f9918c46429ac97bcb51885c76c62558bcc)

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java (hunk 10 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java:2178]: IMO It doesn't make sense to test *patterns* at all for either `AuthorizedIndices#all` or `AuthorizedIndices#check`. In the main code it's only names of resources that are passed as arguments to these methods, never patterns. That's why I removed the test here. The assert can be `true` or `false` depending if the pattern in the request is greater than the the pattern in the permission, but it's not relevant.

Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java:2178]: pushed [b8cd97e](https://github.com/elastic/elasticsearch/pull/92625/commits/b8cd97e51294304dcec8e0734e4adf209d9f20e8)

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/accesscontrol/IndicesPermissionTests.java (hunk 12 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/accesscontrol/IndicesPermissionTests.java]: pushed [b8cd97e](https://github.com/elastic/elasticsearch/pull/92625/commits/b8cd97e51294304dcec8e0734e4adf209d9f20e8)

**Comment by albertzaharovits** [comment]
https://gradle-enterprise.elastic.co/s/73ulifdoml74q tracked by https://github.com/elastic/elasticsearch/issues/91450

**Comment by albertzaharovits** [comment]
@elasticmachine run elasticsearch-ci/part-2

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java (hunk 10 above)
Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java:2178]: > In the main code it's only names of resources that are passed as arguments to these methods, never patterns. 

I think it could happen when `expand_wildcards=none`. Not something we should optimize for. But the `check` should still work correctly for it.

Comment (albertzaharovits) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolverTests.java:2178]: Fair enough, I forgot about that case.
